java-checks/src/main/java/org/sonar/java/checks/LazyArgEvaluationCheck.java

```diff
@@ -0,0 +1,287 @@
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.CatchTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.IfStatementTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.Deque;
+import java.util.List;
+import java.util.Locale;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Rule(key = "S2629")
+public class LazyArgEvaluationCheck extends BaseTreeVisitor implements JavaFileScanner {
+
+  private static final String SLF4J_LOGGER = "org.slf4j.Logger";
+  public static final String JUL_LOGGER = "java.util.logging.Logger";
+  private static final String STRING = "java.lang.String";
+
+  interface LogLevels {
+    List<MethodMatcher> log();
+    MethodMatcher test();
+
+    static MethodMatcher loggerProto(String typeDefinition) {
+      return MethodMatcher.create()
+        .typeDefinition(typeDefinition)
+        .addParameter(STRING);
+    }
+
+    static MethodMatcher levelTestProto(String typeDefinition) {
+      return MethodMatcher.create()
+        .typeDefinition(typeDefinition)
+        .withoutParameter();
+    }
+
+    static Stream<LogLevels> logLevels() {
+      return Stream.concat(Arrays.stream(SLF4J_LEVELS.values()), Arrays.stream(JUL_LEVELS.values()));
+    }
+
+    static List<MethodMatcher> varArgVariants(String logger, String methodName) {
+      return ImmutableList.of(loggerProto(logger).name(methodName),
+        loggerProto(logger).name(methodName).addParameter(TypeCriteria.anyType()),
+        loggerProto(logger).name(methodName).addParameter(TypeCriteria.anyType()).addParameter(TypeCriteria.anyType()));
+    }
+
+    enum SLF4J_LEVELS implements LogLevels {
+      TRACE,
+      DEBUG,
+      INFO,
+      WARN,
+      ERROR;
+
+      final List<MethodMatcher> log = varArgVariants(SLF4J_LOGGER, toString().toLowerCase(Locale.ROOT));
+
+      final MethodMatcher test = levelTestProto(SLF4J_LOGGER).name(String.format("is%c%sEnabled", toString().charAt(0), toString().toLowerCase(Locale.ROOT).substring(2)));
+
+      @Override
+      public List<MethodMatcher> log() {
+        return log;
+      }
+
+      @Override
+      public MethodMatcher test() {
+        return test;
+      }
+    }
+
+    enum JUL_LEVELS implements LogLevels {
+      SEVERE,
+      WARNING,
+      INFO,
+      CONFIG,
+      FINE,
+      FINER,
+      FINEST;
+
+      final List<MethodMatcher> log = ImmutableList.of(loggerProto(JUL_LOGGER).name(toString().toLowerCase(Locale.ROOT)));
```

Comment:
I'm not convinced about the usage of the log/test fields in both enums. I wonder if they are really useful. From what I can see, the `log()` and `test()` methods will be called only once, in the static initializers of the rule, so maybe we can win some place and simply put their initializers directly in the implementations of the methods. It won't make it less clear.